daemon/cdi_linux.go

```diff
@@ -0,0 +1,95 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	"github.com/docker/docker/pkg/capabilities"
+	"github.com/hashicorp/go-multierror"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// cdiDeviceInjector defines the operations required to inject CDI devices into an OCI spec.
+type cdiDeviceInjector interface {
+	InjectDevices(*specs.Spec, ...string) ([]string, error)
+	GetErrors() map[string][]error
+}
+
+type cdiHandler struct {
+	registry cdiDeviceInjector
+}
+
+// RegisterCDIDriver registers the CDI device driver.
+// The driver injects CDI devices into an incoming OCI spec and is called for DeviceRequests associated with CDI devices.
+func RegisterCDIDriver(opts ...cdi.Option) {
+	cache, err := cdi.NewCache(opts...)
+	if err != nil {
+		// We create a spec updater that always returns an error.
+		// This error will be returned only when a CDI device is requested.
+		// This ensures that daemon startup is not blocked by a CDI registry initialization failure.
+		errorOnUpdateSpec := func(s *specs.Spec, dev *deviceInstance) error {
+			return fmt.Errorf("CDI device injection failed due to registry initialization failure: %w", err)
+		}
+		driver := &deviceDriver{
+			capset:     capabilities.Set{"cdi": struct{}{}},
+			updateSpec: errorOnUpdateSpec,
+		}
+		registerDeviceDriver("cdi", driver)
+		return
+	}
+
+	// We construct a spec updates that injects CDI devices into the OCI spec using the initialized registry.
+	c := &cdiHandler{
+		registry: cache,
+	}
+
+	driver := &deviceDriver{
+		capset:     capabilities.Set{"cdi": struct{}{}},
+		updateSpec: c.injectCDIDevices,
+	}
+
+	registerDeviceDriver("cdi", driver)
+}
+
+// injectCDIDevices injects a set of CDI devices into the specified OCI specification.
+func (c *cdiHandler) injectCDIDevices(s *specs.Spec, dev *deviceInstance) error {
+	if dev == nil {
+		return nil
+	}
```

Comment:
Should this be an error? There's no docs on this, but I can't see this ever being a sensible thing to happen for this callback. We don't check `s` (but maybe `InjectDevices` handles a nil spec?); the nvidia driver will nil-dereference panic if either is `nil`.

Reply:
I agree; it wouldn't ever make sense for this callback—or any `deviceDriver.updateSpec` callback—to be called with either argument `nil`. I'd go so far as to consider it a bug in the caller if it did so. The call definitely should fail loudly, not silently. I would be perfectly happy if you simply let it nil-dereference panic like the nvidia driver does since this will be called from an HTTP request-handler goroutine. A goroutine stack trace would be written to the daemon log, which would be more helpful for debugging than returning an error to the Engine API client.